### PR TITLE
GH-118093: Add tier two support to several instructions

### PR DESCRIPTION
**Pull request by brandtbucher**
This adds tier two support to all of the the *trivial* instructions that our stats indicate are causing traces to end (`CALL_LIST_APPEND`, `IMPORT_NAME`, `LOAD_NAME`, `BUILD_SET`, `SEND_GEN`, and `IMPORT_FROM`). Some other, more complicated, instructions will come in their own PRs. It also turns `_FOR_ITER_TIER_TWO` from a deopting instruction into an exiting instruction, since we expect it to fail fairly often (since this represents normal control flow in the program).

Perf is [neutral](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20240715-3.14.0a0-6f4aaba-JIT/bm-20240715-linux-x86_64-brandtbucher-tier_two_improvement-3.14.0a0-6f4aaba-vs-base.svg). Stats have [improved](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20240715-3.14.0a0-6f4aaba-JIT/bm-20240715-azure-x86_64-brandtbucher-tier_two_improvement-3.14.0a0-6f4aaba-pystats-vs-base.md), as expected (4% fewer tier one instructions, 24% fewer too-short traces, 20% fewer optimization attempts, 7% fewer traces executed, 4% more traces created, and 4% more uops executed).

* Issue: gh-118093

### CODE REVIEW

File: Python/bytecodes.c [hunk 1]
```diff
@@ -3955,16 +3953,15 @@ dummy_func(
...
             STAT_INC(CALL, hit);
-            if (_PyList_AppendTakeRef((PyListObject *)self_o, PyStackRef_AsPyObjectSteal(arg)) < 0) {
-                goto pop_1_error;  // Since arg is DECREF'ed already
-            }
+            int err = _PyList_AppendTakeRef((PyListObject *)self_o, PyStackRef_AsPyObjectSteal(arg));
             PyStackRef_CLOSE(self);
             PyStackRef_CLOSE(callable);
-            STACK_SHRINK(3);
+            ERROR_IF(err, error);
+        #if TIER_ONE
```
Comment (markshannon) [inline Python/bytecodes.c:3957]: At first glance, this appears incorrect. The code should either push `None` *or* skip the `POP_TOP`.
Can you add a comment here that the skip is handed for tier 2 during trace projection.

**Review by markshannon — CHANGES_REQUESTED** [review]
One minor clarification needed, otherwise LGTM.

Regarding handling `CALL_LIST_APPEND` in tier 2:
Instead of asserting that the next instruction is `POP_TOP` and skipping it, we could emit a `LOAD_CONST_BORROW None`.
If the following instruction is not `POP_TOP` then then code is still correct. If there is a `POP_TOP` following, then optimizer will eliminate the `LOAD; POP_TOP`.

**Comment by brandtbucher** [comment]
I have made the requested changes; please review again

**Comment by brandtbucher** [comment]
JIT failures are unrelated: GH-121946

**Pull request merged**